### PR TITLE
Ssl certificates

### PR DESCRIPTION
**Pull request by LupascuAndrei**
Self signed generation from simba. Tested its functionality from python on pumbaa.

Includes all changes from https://github.com/eerimoq/simba/pull/115 

Since OTA updates are a thing, I don't think it's safe to download a new binary which contains the certificate & key for ssl. Some REST APIs ( like postman ) require the host to have certificates, even if they are self signed and untrusted. This is the best workaround i could come up with; these changes provide you the possibility to generate a new (cert, key) pair of size 1024 ( i would not recommend generating keys with dimensions larger than that because of the time it takes on esp32 ). 

examples i've followed :
https://tls.mbed.org/api/cert__write_8c_source.html
https://tls.mbed.org/api/gen__key_8c_source.html 

used m_malloc instead as recommended here to not be needed to increasee simba stack size 
 https://github.com/eerimoq/pumbaa/issues/35 ; 
the implementation supports other key types generation (i.e. rsa / ca ) , but i only made stress tests on the default configuration.

Does it need to be a separate module than ssl, since its functionality is useless if not used with the ssl module?

### CODE REVIEW

File: 3pp/compat/mbedtls/library/mbedtls_x509write_crt.c [hunk 1]
```diff
@@ -320,32 +324,32 @@ int mbedtls_x509write_crt_der( mbedtls_x509write_cert *ctx, unsigned char *buf,
...
+        return( val );
     }
 
     /*
      *  Extensions  ::=  SEQUENCE SIZE (1..MAX) OF Extension
      */
-    MBEDTLS_ASN1_CHK_ADD( len, mbedtls_x509_write_extensions( &c, tmp_buf, ctx->extensions ) );
-    MBEDTLS_ASN1_CHK_ADD( len, mbedtls_asn1_write_len( &c, tmp_buf, len ) );
-    MBEDTLS_ASN1_CHK_ADD( len, mbedtls_asn1_write_tag( &c, tmp_buf, MBEDTLS_ASN1_CONSTRUCTED |
+    MBEDTLS_ASN1_CHK_ADD2( len, mbedtls_x509_write_extensions( &c, tmp_buf, ctx->extensions ) );
```
Comment (eerimoq) [inline 3pp/compat/mbedtls/library/mbedtls_x509write_crt.c]: Why is the ADD2 macro used instead of ADD?

File: 3pp/compat/mbedtls/library/mbedtls_x509write_crt.c [hunk 2]
```diff
@@ -403,14 +407,14 @@ int mbedtls_x509write_crt_der( mbedtls_x509write_cert *ctx, unsigned char *buf,
     if( ( ret = mbedtls_pk_sign( ctx->issuer_key, ctx->md_alg, hash, 0, sig, &sig_len,
                          f_rng, p_rng ) ) != 0 )
     {
-        return( ret );
+        return( -2 );
```
Comment (eerimoq) [inline 3pp/compat/mbedtls/library/mbedtls_x509write_crt.c]: Why is this changed?

File: 3pp/compat/mbedtls/user_config.h [hunk 3]
```diff
@@ -2,9 +2,9 @@
  * @section License
  *
  * The MIT License (MIT)
- *
- * Copyright (c) 2014-2017, Erik Moqvist
- *
+ * 
+ * Copyright (c) 2014-2016, Erik Moqvist
+ *
```
Comment (eerimoq) [inline 3pp/compat/mbedtls/user_config.h]: Why change the year?

File: 3pp/compat/mbedtls/user_config.h [hunk 4]
```diff
@@ -35,16 +35,16 @@
...
 #undef MBEDTLS_SSL_DTLS_HELLO_VERIFY
 #undef MBEDTLS_SSL_DTLS_CLIENT_PORT_REUSE
 #undef MBEDTLS_SSL_DTLS_ANTI_REPLAY
 #undef MBEDTLS_SSL_DTLS_BADMAC_LIMIT
 
-#undef MBEDTLS_DEBUG_C
+//#undef MBEDTLS_DEBUG_C
 #undef MBEDTLS_BLOWFISH_C
-#undef MBEDTLS_X509_CSR_WRITE_C
+//#undef MBEDTLS_X509_CSR_WRITE_C
```
Comment (eerimoq) [inline 3pp/compat/mbedtls/user_config.h]: Do we want those features to be enabled?

File: src/inet/ssl.c [hunk 5]
```diff
@@ -30,7 +30,9 @@
 
 #include "simba.h"
 
+
 #include "mbedtls/config.h"
+
```
Comment (eerimoq) [inline src/inet/ssl.c]: Why adding a blank line?

File: src/inet/ssl.c [hunk 6]
```diff
@@ -41,53 +43,91 @@
...
+            mbedtls_x509_crt_free(&module.cert);
+    if(module.allocated[SSL_MODULE_ALLOCATED_KEY])
+            mbedtls_pk_free(&module.key);
+    if(module.allocated[SSL_MODULE_ALLOCATED_CA_CERTS])
+            mbedtls_x509_crt_free(&module.ca_certs);
+    memset(&module.timer,0,sizeof(module.timer));
+    module.initialized = 0;
+    memset(module.allocated,0,sizeof(module.allocated));
+    ssl_module_init();
+}
```
Comment (eerimoq) [inline src/inet/ssl.c]: Put in destroy function.

File: 3pp/compat/mbedtls/user_config.h (hunk 3 above)
Comment (LupascuAndrei) [inline 3pp/compat/mbedtls/user_config.h]: sorry,typo, i made the pull request over a older version of simba

File: 3pp/compat/mbedtls/library/mbedtls_x509write_crt.c (hunk 1 above)
Comment (LupascuAndrei) [inline 3pp/compat/mbedtls/library/mbedtls_x509write_crt.c]: instead of returning "-1" regardless of the error encountered, i'm returning an integer which represesnts "CHK_ADD2" line where it crashed.
used me a lot during debugging, i consider it stable now, should i revert to ADD instead of ADD2?

Comment (eerimoq) [inline 3pp/compat/mbedtls/library/mbedtls_x509write_crt.c]: Yeah, the fewer changes in the code the better. Makes it easier to make an uplift of mbedtls when required.

**Comment by LupascuAndrei** [comment]
will rewrite some parts of it and make a pull request soon
